### ISSUE TITLE
How to test on custom dataset

### ISSUE THREAD

**Issue by krutikabapat** [malllabiisc/DiPS#15]
@ashutoshml I am traying to make this run for my custom dataset. I have some questions, which I have kept in the src.txt file. I wanted to know what should I keep in the tgt.txt file. Thanks in advance for the help :D

**Comment by ashutoshml** [comment]
Hi @krutikabapat 

The tgt.txt file during test is just used for evaluation, a file that has gold paraphrases of the original sentences. Since you don't necessarily require evaluation, you can just copy the same src.txt file, rename it to tgt.txt and it should work. 

Let me know in case you still face any issues.

**Comment by krutikabapat** [comment]
@ashutoshml Thanks a lot for the help. I have one more doubt. So for my own custom dataset. Can I keep the train folder as it is(quora dataset) and keep my own dataset in the train and evaluation set, keeping the same content (in the src.txt and tgt.txt file)? Will it make sense?
So it would be like Train(quora - src.txt, tgt.txt (both files different)
Test (custom dataset - src.txt, tgt.txt (both files same))
Validation(custom dataset -src.txt, tgt.txt(both files same))

**Comment by ashutoshml** [comment]
This setting should work.

However, it would be helpful if you could give a better validation set (probably an exclusive set from the training data) because that ultimately helps in deciding which checkpoint works best and is worth saving. That is to say, model selection is done through the validation dataset.

**Comment by ashutoshml** [comment]
Closing the issues. 
Please re-open if you face the problem again.

**Comment by krutikabapat** [comment]
@ashutoshml When I try to generate questions on my own dataset, as mentioned above it shoqs results like below:-
'''
2020-05-28 03:42:15,380 | INFO | __main__ | Batch: 6800	Loss: 1.4282072067260743
2020-05-28 03:43:08,199 | INFO | __main__ | Batch: 7000	Loss: 1.4196510314941406
2020-05-28 03:44:00,997 | INFO | __main__ | Batch: 7200	Loss: 1.7451730728149415
2020-05-28 03:44:53,923 | INFO | __main__ | Batch: 7400	Loss: 1.347912883758545
2020-05-28 03:45:20,606 | INFO | __main__ | Sample Generations
2020-05-28 03:45:20,607 | INFO | __main__ | ==========================================
2020-05-28 03:45:20,607 | INFO | __main__ | ==========================================
2020-05-28 03:45:20,607 | INFO | __main__ | ==========================================
2020-05-28 03:45:20,747 | INFO | __main__ | ==========================================
2020-05-28 03:45:20,747 | INFO | __main__ | ==========================================
2020-05-28 03:45:20,747 | INFO | __main__ | ==========================================
2020-05-28 03:45:20,748 | INFO | __main__ | Epoch: 2	BLEU: (0.0, [0.15384615384615385, 0.0, 0.0, 0.0], 1.0, 1.1818181818181819, 13, 11)
2020-05-28 03:45:20,749 | INFO | __main__ | Epoch: 2	Train loss: 12262.711601621293	Val loss: 9.004253387451172
2020-05-28 03:45:20,749 | INFO | __main__ | Epoch: 3
2020-05-28 03:46:13,649 | INFO | __main__ | Batch: 200	Loss: 1.3494608879089356
2020-05-28 03:47:06,865 | INFO | __main__ | Batch: 400	Loss: 1.0824223518371583
2020-05-28 03:47:59,912 | INFO | __main__ | Batch: 600	Loss: 1.2722644805908203
2020-05-28 03:48:52,895 | INFO | __main__ | Batch: 800	Loss: 1.120955753326416
2020-05-28 03:49:45,897 | INFO | __main__ | Batch: 1000	Loss: 1.694565200805664
2020-05-28 03:50:38,803 | INFO | __main__ | Batch: 1200	Loss: 3.443445587158203
2020-05-28 03:51:31,675 | INFO | __main__ | Batch: 1400	Loss: 1.126705265045166
'''
Basically the generations are blank. Please help me with this :D

**Comment by ashutoshml** [comment]
Hi @krutikabapat 

I am re-opening the issue.

Please let me know your exact steps to reproduce this. 
I'll try to run It on my system and get back to you by the end of the first week of June.

If you figure out the issue yourself by then, I would appreciate it if you can update this thread with the relevant issue and its fix.

**Comment by krutikabapat** [comment]
@ashutoshml Sorry for late reply. I thought if i could rectify it myself, but I could not. Please help me with that. I have followed the following steps:
1. Kept the training and validation dataset same and then added some random questions in the test folder.
2. Run this command ```python -m src.main -mode train -gpu 0 -use_attn -bidirectional -dataset quora -run_name ```
And then gout the output above.

**Comment by ashutoshml** [comment]
@krutikabapat: The generations that you get during training are of the validation dataset. During training, the test set remains untouched. See this [line](https://github.com/malllabiisc/DiPS/blob/8ea9747b1c104b6490000cdff15a522e9b860c05/src/main.py#L27)

Only during decoding is the test folder accessed. I am hoping the code has not been modified during training to output generations for the test folder since it doesn't get "loaded" during training.

**Comment by krutikabapat** [comment]
@ashutoshml Thanks a lot for your help, but I haven't changed anything (I mean codewise)

**Comment by ashutoshml** [comment]
@krutikabapat I am guessing you skipped the run_name parameter

`python -m src.main -mode train -gpu 0 -use_attn -bidirectional -dataset quora -run_name`

Please use
`python -m src.main -mode train -gpu 0 -use_attn -bidirectional -dataset quora -run_name quorarun`

**Comment by krutikabapat** [comment]
@ashutoshml Thank you, previously I was using the run_name as testrun. I will try this out.

**Comment by ashutoshml** [comment]
@krutikabapat Any run_name would work well (It should just have alphanumeric characters). In your previous [comment](https://github.com/malllabiisc/DiPS/issues/15#issuecomment-638055755) you did not mention the run_name, which made me think that missing run name might be an issue.

If you have provided testrun as run_name, then this is not the issue.

I am unable to reproduce the issue based on your comments.
 
Can you do a fresh install of everything and check with the default dataset?
That should work, ideally. 

You can later add in your custom test set once the training is complete. It's actually surprising because the test set is completely untouched during training.

**Comment by ashutoshml** [comment]
@krutikabapat Any updates on this?

**Comment by krutikabapat** [comment]
@ashutoshml  I am trying to train that again. I had a one doubt that when I run
```python -m src.main -mode decode -selec submod -run_name <run_name> -beam_width 10 -gpu 0```
What has to be given inside ``` selec submod```.

**Comment by krutikabapat** [comment]
@ashutoshml Also how to train it in CPU if thw GPU is not available? I mean the argument `gpu`?

**Comment by ashutoshml** [comment]
@krutikabapat I don't follow your first question. -selec is a parameter and submod is the argument for that parameter.
What do you mean when you say "inside"?

This code has been written with GPU in mind, it wouldn't directly work with CPU. However, it should not be too difficult to modify to CPU though.

**Comment by krutikabapat** [comment]
@ashutoshml Thank you. For the first one, my bad I got it. And for the CPU, can u suggest some changes in the code.

**Comment by ashutoshml** [comment]
I believe changing this [line](https://github.com/malllabiisc/DiPS/blob/8ea9747b1c104b6490000cdff15a522e9b860c05/src/args.py#L37) to `parser.add_argument('-gpu', type=str, default=None, help='Specify the gpu to use')` should be the first change. 

And use `
python -m src.main -mode train -use_attn -bidirectional -dataset quora -run_name <run_name>` Notice that I have excluded the -gpu parameter here.

You might also need to modify [this function](https://github.com/malllabiisc/DiPS/blob/8ea9747b1c104b6490000cdff15a522e9b860c05/src/helper.py#L123), if need be.

**Comment by krutikabapat** [comment]
@ashutoshml Okay great. Thanks a lot
